File: array/arrayChooseK.js

// unit tests addElementToArray
// addElementToArray(1, [[1],[2]]);
function addElementToArray(value, array) {
  for (var i = 0; i < array.length; i++) {
    array[i].push(value);
  }
  return array;
}

// unit test: removeEntryAleadyWithK
// var answer = [];
// var test = removeEntryAleadyWithK(3, [[1,2,3], [1,2], [1,3]], answer);
// console.log(test);
// console.log(answer);
function removeEntryAleadyWithK(k, array, answer) {
  var newArray = [];
  for (var i = 0; i < array.length; i++) {
    if (array[i].length === k) {
      answer.push(array[i]);
    } else {
      newArray.push(array[i]);
    }
  }
  return newArray;
}

function arrayChooseK(k, array) {
  var answer = [];
  var tmpStore = [[]];

  for (var i = 0; i < array.length; i++) {
    var cloneTmp = tmpStore.map(elem => [...elem]); // deep copy of inner array as well
    cloneTmp = addElementToArray(a[i], cloneTmp);
    var newTmp = removeEntryAleadyWithK(k, cloneTmp, answer);
    tmpStore = [...newTmp, ...tmpStore];
  }

  return answer;
}

// main test
// node arrayChooseK.js

var a = [1, 2, 3, 4];
var k = 3;
console.log(arrayChooseK(k, a)); // [ [ 1, 2, 3 ], [ 2, 3, 4 ], [ 1, 3, 4 ], [ 1, 2, 4 ] ]

File: array/arrayChooseK.py
import copy

def addElementToArray(ele, array):
	for x in array:
  		x.append(ele)
# unit tests addElementToArray
# mylist = [[1],[2]] #[[]];
# addElementToArray( 3, mylist );
# print "Values outside the function: ", mylist


def removeEntryAleadyWithK(k, array, answer):
	newArray = []
	for x in array:
		if (len(x) == k):
			answer.append(x)
		else:
  			newArray.append(x)

	return newArray
# unit test for removeEntryAlreadyWithK
# answer = [];
# remainSet = removeEntryAleadyWithK(3, [[1,2,3], [1,2], [1,3]], answer);
# print "answer: ", answer
# print "remainSet: ", remainSet

def arrayChooseK(k, array):
	answer = []
	tmpStore = [[]]
	for x in array:
		cloneTmp = copy.deepcopy(tmpStore)
		addElementToArray(x, cloneTmp)
		newTmp = removeEntryAleadyWithK(k, cloneTmp, answer)
		tmpStore = newTmp + tmpStore
	return answer

a = [1, 2, 3, 4]
k = 3
answer = arrayChooseK(k, a)
print "array choose 3: ", answer # [ [ 1, 2, 3 ], [ 2, 3, 4 ], [ 1, 3, 4 ], [ 1, 2, 4 ] ]
